fog/view/server/src/fog_view_shard_responses_processor.rs

```diff
@@ -0,0 +1,236 @@
+
+    for (shard_client, mut response) in clients_and_responses {
+        // We did not receive a query_response for this shard.Therefore, we need to:
+        //  (a) retry the query
+        //  (b) authenticate with the Fog View Store that returned the decryption_error
+        if response.has_decryption_error() {
+            shard_clients_for_retry.push(shard_client);
+            let store_uri =
+                FogViewUri::from_str(&response.get_decryption_error().fog_view_store_uri)?;
+            view_store_uris_for_authentication.push(store_uri);
+        } else {
+            new_query_responses.push(response.take_query_response());
+        }
+    }
+
+    Ok(ProcessedShardResponseData::new(
+        shard_clients_for_retry,
+        view_store_uris_for_authentication,
+        new_query_responses,
+    ))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use grpcio::ChannelBuilder;
+    use mc_common::logger::{test_with_logger, Logger};
+    use mc_util_grpc::ConnectionUriGrpcioChannel;
+
+    fn create_successful_mvq_response() -> MultiViewStoreQueryResponse {
+        let mut successful_response = MultiViewStoreQueryResponse::new();
+        let client_auth_request = vec![].into();
+        successful_response
+            .mut_query_response()
+            .set_data(client_auth_request);
+
+        successful_response
+    }
+
+    fn create_failed_mvq_response(i: usize) -> MultiViewStoreQueryResponse {
+        let mut failed_response = MultiViewStoreQueryResponse::new();
+        let view_uri_string = format!("insecure-fog-view://node{}.test.mobilecoin.com:3225", i);
+        failed_response
+            .mut_decryption_error()
+            .set_fog_view_store_uri(view_uri_string);
+
+        failed_response
+    }
+
+    fn create_grpc_client(i: usize, logger: Logger) -> Arc<FogViewApiClient> {
+        let view_uri_string = format!("insecure-fog-view://node{}.test.mobilecoin.com:3225", i);
+        let view_uri = FogViewUri::from_str(&view_uri_string).unwrap();
+        let grpc_env = Arc::new(
+            grpcio::EnvBuilder::new()
+                .name_prefix("Main-RPC".to_string())
+                .build(),
+        );
+
+        let grpc_client = FogViewApiClient::new(
+            ChannelBuilder::default_channel_builder(grpc_env.clone())
+                .connect_to_uri(&view_uri, &logger),
+        );
+
+        Arc::new(grpc_client)
+    }
+
+    #[test_with_logger]
+    fn one_successful_response_no_shard_clients(logger: Logger) {
+        let grpc_client = create_grpc_client(0, logger.clone());
+        let successful_mvq_response = create_successful_mvq_response();
+        let clients_and_responses = vec![(grpc_client, successful_mvq_response)];
+
+        let result = process_shard_responses(clients_and_responses);
+
+        assert!(result.is_ok());
+
+        let shard_clients_for_retry = result.unwrap().shard_clients_for_retry;
+        assert!(shard_clients_for_retry.is_empty());
```

Comment:
While `Vec<Arc<_>>` can't derive `PartialEq` would it still be good to write the tests like it was comparing to a `ProcessedShardResponseData`.
as it currently stands there is a bit of repetition (which tests may have) and it can be hard to spot differences.
```suggestion
        let responses = process_shard_responses(clients_and_responses).unwrap();

        assert!(responses.shard_clients_for_retry.is_empty());
        assert!(responses.view_store_uris_for_authentication.is_empty());
        assert_eq!(responses.new_query_responses.len(), 1);
```

Reply:
I tend to write tests such that there is one `assert` statement per test. This makes it easier for people who break the tests to understand what is going wrong. When a test has multiple assert statements they'll just see that the test name is broken, and the test name normally isn't descriptive enough to indiciate the specific failure. I name the tests such that if they fail then the reader understands what went wrong.

 I'm going to leave it as is unless you feel strongly, feel free to let me know.

Reply:
I think that's fine, especially for unit tests one assert per test is ideal, and I would like to see more of that versus some of the bigger recipe tests I've seen in the repo.

Reply:
Yeah, I agree that the bigger recipe tests should be broken down into smaller ones.